pkg/meta/openfile.go

```diff
@@ -20,36 +20,68 @@ type openFile struct {
 
 type openfiles struct {
 	sync.Mutex
-	expire time.Duration
-	files  map[Ino]*openFile
+	expire  time.Duration
+	limit   uint64
+	timeout time.Duration
+	files   map[Ino]*openFile
 }
 
-func newOpenFiles(expire time.Duration) *openfiles {
+func newOpenFiles(expire time.Duration, limit uint64, timeout time.Duration) *openfiles {
+	if limit == 0 {
+		limit = 1e4
+	}
+	if timeout == 0 {
+		timeout = 12 * time.Hour
+	}
 	of := &openfiles{
-		expire: expire,
-		files:  make(map[Ino]*openFile),
+		expire:  expire,
+		limit:   limit,
+		timeout: timeout,
+		files:   make(map[Ino]*openFile),
 	}
 	go of.cleanup()
 	return of
 }
 
 func (o *openfiles) cleanup() {
 	for {
+		unusedFiles := make(map[Ino]*openFile)
 		o.Lock()
-		cutoff := time.Now().Add(-time.Hour).Add(time.Second * time.Duration(len(o.files)/1e4))
-		var cnt, expired int
+		var cnt, deleted int
 		for ino, of := range o.files {
-			if of.refs <= 0 && of.lastCheck.Before(cutoff) {
-				delete(o.files, ino)
-				expired++
-			}
 			cnt++
 			if cnt > 1e3 {
 				break
 			}
+			if of.refs <= 0 {
+				unusedFiles[ino] = of
```

Comment:
we can delete the file here rather than save it in a map

Reply:
done

Reply:
done